source/scpd/quorum/QuorumTracker.d

```diff
@@ -0,0 +1,67 @@
+/*******************************************************************************
+
+    Bindings for quorum/QuorumTracker.h
+
+    Note: D code does not need to use anything in this class except the
+    QuorumMap definition.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module scpd.quorum.QuorumTracker;
+
+@trusted @nogc nothrow:
+
+import scpd.Cpp;
+import scpd.scp.SCP;
+import scpd.types.Stellar_SCP;
+import scpd.types.Stellar_types;
+
+extern (C++, `stellar`):
+
+/// helper class to help track the overall quorum over time
+/// a node tracked is definitely in the transitive quorum.
+/// If its associated quorum set is empty (nullptr), it just means
+/// that another node has that node in its quorum set
+/// but could not explore the quorum further (as we're missing the quorum set)
+/// Nodes can be added one by one (calling `expand`, most efficient)
+/// or the quorum can be rebuilt from scratch by using a lookup function
+extern(C++, class) public class QuorumTracker
+{
+public:
+    /// The Node ID => Quorum set map type
+    alias QuorumMap = unordered_map!(NodeID, SCPQuorumSetPtr);
+
+private:
+    SCP* mSCP;
+    QuorumMap mQuorum;
+
+public:
+    /// Initialize a new QuorumTracker
+    this (SCP* scp);
+
+    /// returns true if id is in transitive quorum for sure
+    bool isNodeDefinitelyInQuorum(const ref NodeID id);
```

Comment:
> In D those functions are virtual

Oh damn. How come it worked?

> From my experience, things declared as class with no virtual method => struct.

What about classes which inherit other classes? This one specifically is defined as `class QuorumTracker : public NonMovableOrCopyable`, so it looks like it was never supposed to be used as a POD type.

Reply:
> class QuorumTracker : public NonMovableOrCopyable

Sounds like the idiomatic way to do this in D would be:
```D
mixin template NonMovableOrCopyable (T /* = typeof(this) */) // Not sure if the default arg works
{
    @disable this(this);
    @disable ref T opAssign () (auto ref T other);
}

public struct QuorumTracker
{
    mixin NonMovableOrCopyable!(QuorumTracker);
}
```

That might be a bit of an overreaching statement, but my experience is that good, principled C++ code ends up being idiomatic D code, and trying to use the same construct in D as one does in C++ might lead one astray.

> Oh damn. How come it worked?

Depends on the usage. The code called from C++ doesn't have an issue, obviously. The code called from D may or may not crash, depending on whether or not the compiler devirtualize the call. And since we dropped DMD, there is a good chance it does :)

Reply:
Note, the above snippet doesn't fully capture the C++ intent though, since in D, you *cannot* declare a `struct` are non movable. Should probably include a copy constructor too but it currently triggers a cascade of bugs.

Reply:
> And since we dropped DMD, there is a good chance it does :)

LOL.

Yeah also I dislike virtual by default in D. Where is Manu when you need him.